### DISCUSSION TITLE
hydrating goes wrong | webpack | babel-loader | babel-preset-solid

### DISCUSSION THREAD

**Discussion by BeeckmanThe1** [solidjs/solid#1857] [Q&A]
I have this component:

```
import {createEffect, createSignal} from "solid-js";

const HelloWorld = () => {
    createEffect(() => {
        console.log('lel!')
    })
    const [getValue, setValue] = createSignal(0)

    return <div>
        <h1>Hello</h1>
        <button onclick={() => alert('lal')}>Click me {getValue()}</button>
    </div>
}

export default HelloWorld
```

It gets server side rendered:

```
export const getBasicHTMLSkeleton = () => {
    const hello = renderToString(() => <HelloWorld/>)
    return `
  <html lang="en">
    <head>
      <title>🔥 Solid SSR 🔥</title>
      <meta charset="UTF-8" />
      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="stylesheet" href="/_index.min.css" />
      <script defer src="hydrate.js" />
      ${generateHydrationScript()}
    </head>
    <body>
        <div id="root">
            ${hello}
        </div>
    </body>
  </html>
`;
}
```

This is the hydrate.tsx script I use:

```
import {hydrate} from "solid-js/web";
import HelloWorld from "../hybrid/components/HelloWorld";

const hydrateSections = () => {
    try {
            const rootNode = document.getElementById('root')
            console.log('rootNode', rootNode)
            console.log('HYDRATING?')
            hydrate(() => <HelloWorld/>, rootNode)
    } catch (err) {
        console.log('err', err)
        debugger
    }
}

window.onload = () => hydrateSections()
```

It gets compiled using this webpack-babel setup:

```
const configClient = {
    devtool: !isProduction && 'source-map',
    mode: isProduction ? 'production' : 'development',
    resolve: {
        extensions: ['.tsx', '.ts', '.jsx', '.js']
    },
    target: 'web',
    stats: 'errors-only',
    entry: './src/client/hydrate.tsx',
    module: {
        rules: [{
            test: /\.tsx?$/,
            exclude: /node_modules/,
            use: [
                {
                    loader: 'babel-loader',
                    options: {
                        presets: [['solid', { "generate": "dom", "hydratable": true }]],
                    },
                },
                {
                    loader: 'ts-loader',
                },
            ]
        },
            filesRulesClient
        ]
    },
    output: {
        filename: 'hydrate.js',
        path: path.resolve(__dirname, 'dist-client')
    }
}
```
I am following this documentation quite strict: https://www.npmjs.com/package/babel-preset-solid.

When the page load & executes the hydrate script, I get this error:

```
err TypeError: Cannot read properties of undefined (reading 'completed')
    at hydrate$1 (dev.js:256:25)
    at hydrate (dev.js:565:1)
    at hydrateSections (hydrate.tsx:9:20)
    at window.onload (hydrate.tsx:16:23)
```

The rest seems to be working:
![image](https://github.com/solidjs/solid/assets/59286357/c68c78b4-bbe7-4e74-9397-488f9c0e2531)

Did I miss something?

**Comment by BeeckmanThe1** [comment]
I got it working after a few days.
<script defer src="hydrate.js" /> should be: 👉 <script async src="hydrate.js" />

I think there is some opportunity to improve the documentation here.
eg: https://www.solidjs.com/guides/server: 

![image](https://github.com/solidjs/solid/assets/59286357/b038f66a-a02b-4e50-99ef-dddf1ac02240)


This part is just empty: https://docs.solidjs.com/references/concepts/ssr/hydration

I would do it myself, but I am brand new to Solidjs so it would be a bit weird to suggest documentation improvements.

**Reply by snnsnn** [reply]
Not necessarily true. Hydration requires the DOM elements being rendered first, only then they can be hydrated. If you call the hydrate method before elements are rendered, you will have error like the one you had because you will be trying to style and manipulate non-existent elements.

**Reply by BeeckmanThe1** [reply]
Yeah, that's why you need to add the `async` attribute, right?

I think it should be mentioned in the docs that the `<HydrationScript/>` (or the `generateHydrationScript()`) only does the preparation for the hydration and that you need to import the script where the actual hydration happens yourself and that this second script needs the `async` attribute. This is not straight forward, but something that should be mentioned in the documentation.
Also, what is the reason the `<HydrationScript/>` is not doing the hydrating after everything is loaded?

**Reply by snnsnn** [reply]
> what is the reason the `<HydrationScript/>` is not doing the hydrating after everything is loaded?

It inserts script elements, `<script>`, that rebuild the application on the client side.
